### ISSUE TITLE
Output shape of ONNX exported "index_select" function is wrong on pytorch 1.10.0

### ISSUE THREAD

**Issue by gnloch** [pytorch/pytorch#68003]
## 🐛 Bug

When trying to export a model to ONNX that uses the `index_select` function on pytorch 1.10, the resulting operation in ONNX doesn't have the expected shape:

## To Reproduce

```python
import torch
from torch import nn


class IndexSelectModule(nn.Module):
    def __init__(self):
        super().__init__()
        self.register_buffer('channel_order', torch.tensor([2, 1, 0]))

    def forward(self, x):
        # x1 = x[:, [2, 1, 0], :, :]
        x1 = torch.index_select(x, 1, self.channel_order)
        return x1


class MaxPoolModule(nn.Module):
    def __init__(self):
        super().__init__()
        self.register_buffer('channel_order', torch.tensor([2, 1, 0]))
        self.avgpool = nn.AdaptiveMaxPool2d((4, 4))

    def forward(self, x):
        # x1 = x[:, [2, 1, 0], :, :]
        x1 = torch.index_select(x, 1, self.channel_order)
        x2 = self.avgpool(x1)
        return x2


def to_onnx(m, name):
    dummy = torch.rand((2, 3, 16, 16))
    with torch.no_grad():
        torch.onnx.export(m, dummy,
                          f'{name}.pytorch-{torch.__version__}.onnx',
                          input_names=['input'],
                          output_names=['output'],
                          do_constant_folding=True,
                          dynamic_axes={'input': {0: 'batch'}, 'output': {0: 'batch'}})



def main():
    m1 = IndexSelectModule()
    m2 = MaxPoolModule()
    m1.eval()
    m2.eval()

    to_onnx(m1, 'index_select')
    to_onnx(m2, 'max_pool')


if __name__ == '__main__':
    main()

```
1. Run snippet using pytorch 1.8.1
2. Run again using pytorch 1.10.0

Output using pytorch 1.10:
```sh
(torch1100_py38) [gloch@machine bug_test]$ python bug_test.py 
/home/gloch/.miniconda3/envs/torch1100_py38/lib/python3.8/site-packages/torch/onnx/symbolic_helper.py:258: UserWarning: ONNX export failed on adaptive_max_pool2d because input size not accessible not supported
  warnings.warn("ONNX export failed on " + op + " because " + msg + " not supported")
WARNING: The shape inference of aten::adaptive_max_pool2d type is missing, so it may result in wrong shape inference for the exported graph. Please consider adding it in symbolic function.
WARNING: The shape inference of aten::adaptive_max_pool2d type is missing, so it may result in wrong shape inference for the exported graph. Please consider adding it in symbolic function.
WARNING: The shape inference of aten::adaptive_max_pool2d type is missing, so it may result in wrong shape inference for the exported graph. Please consider adding it in symbolic function.
WARNING: The shape inference of aten::adaptive_max_pool2d type is missing, so it may result in wrong shape inference for the exported graph. Please consider adding it in symbolic function.
Traceback (most recent call last):
  File "bug_test.py", line 52, in <module>
    main()
  File "bug_test.py", line 48, in main
    to_onnx(m2, 'max_pool')
  File "bug_test.py", line 32, in to_onnx
    torch.onnx.export(m, dummy,
  File "/home/gloch/.miniconda3/envs/torch1100_py38/lib/python3.8/site-packages/torch/onnx/__init__.py", line 316, in export
    return utils.export(model, args, f, export_params, verbose, training,
  File "/home/gloch/.miniconda3/envs/torch1100_py38/lib/python3.8/site-packages/torch/onnx/utils.py", line 107, in export
    _export(model, args, f, export_params, verbose, training, input_names, output_names,
  File "/home/gloch/.miniconda3/envs/torch1100_py38/lib/python3.8/site-packages/torch/onnx/utils.py", line 737, in _export
    proto, export_map, val_use_external_data_format = graph._export_onnx(
RuntimeError: ONNX export failed: Couldn't export operator aten::adaptive_max_pool2d

Defined at:
/home/gloch/.miniconda3/envs/torch1100_py38/lib/python3.8/site-packages/torch/nn/functional.py(1038): adaptive_max_pool2d_with_indices
/home/gloch/.miniconda3/envs/torch1100_py38/lib/python3.8/site-packages/torch/nn/functional.py(1046): _adaptive_max_pool2d
/home/gloch/.miniconda3/envs/torch1100_py38/lib/python3.8/site-packages/torch/_jit_internal.py(422): fn
/home/gloch/.miniconda3/envs/torch1100_py38/lib/python3.8/site-packages/torch/nn/modules/pooling.py(1046): forward
/home/gloch/.miniconda3/envs/torch1100_py38/lib/python3.8/site-packages/torch/nn/modules/module.py(1090): _slow_forward
/home/gloch/.miniconda3/envs/torch1100_py38/lib/python3.8/site-packages/torch/nn/modules/module.py(1102): _call_impl
bug_test.py(25): forward
/home/gloch/.miniconda3/envs/torch1100_py38/lib/python3.8/site-packages/torch/nn/modules/module.py(1090): _slow_forward
/home/gloch/.miniconda3/envs/torch1100_py38/lib/python3.8/site-packages/torch/nn/modules/module.py(1102): _call_impl
/home/gloch/.miniconda3/envs/torch1100_py38/lib/python3.8/site-packages/torch/jit/_trace.py(118): wrapper
/home/gloch/.miniconda3/envs/torch1100_py38/lib/python3.8/site-packages/torch/jit/_trace.py(127): forward
/home/gloch/.miniconda3/envs/torch1100_py38/lib/python3.8/site-packages/torch/nn/modules/module.py(1102): _call_impl
/home/gloch/.miniconda3/envs/torch1100_py38/lib/python3.8/site-packages/torch/jit/_trace.py(1166): _get_trace_graph
/home/gloch/.miniconda3/envs/torch1100_py38/lib/python3.8/site-packages/torch/onnx/utils.py(388): _trace_and_get_graph_from_model
/home/gloch/.miniconda3/envs/torch1100_py38/lib/python3.8/site-packages/torch/onnx/utils.py(437): _create_jit_graph
/home/gloch/.miniconda3/envs/torch1100_py38/lib/python3.8/site-packages/torch/onnx/utils.py(493): _model_to_graph
/home/gloch/.miniconda3/envs/torch1100_py38/lib/python3.8/site-packages/torch/onnx/utils.py(724): _export
/home/gloch/.miniconda3/envs/torch1100_py38/lib/python3.8/site-packages/torch/onnx/utils.py(107): export
/home/gloch/.miniconda3/envs/torch1100_py38/lib/python3.8/site-packages/torch/onnx/__init__.py(316): export
bug_test.py(32): to_onnx
bug_test.py(48): main
bug_test.py(52): <module>


Graph we tried to export:
graph(%input : Float(*, 3, 16, 16, strides=[768, 256, 16, 1], requires_grad=0, device=cpu),
      %channel_order : Long(3, strides=[1], requires_grad=0, device=cpu)):
  %2 : Float(*, *, *, *, strides=[768, 256, 16, 1], requires_grad=0, device=cpu) = onnx::Gather[axis=1](%input, %channel_order) # bug_test.py:24:0
  %3 : Long(2, strides=[1], device=cpu) = onnx::Constant[value= 4  4 [ CPULongType{2} ]]()
  %output : Float(*, *, *, *, strides=[48, 16, 4, 1], requires_grad=0, device=cpu), %5 : Long(*, *, *, *, strides=[48, 16, 4, 1], requires_grad=0, device=cpu) = aten::adaptive_max_pool2d(%2, %3) # /home/gloch/.miniconda3/envs/torch1100_py38/lib/python3.8/site-packages/torch/nn/functional.py:1038:0
  return (%output)
```

![Screenshot from 2021-11-08 12-56-46](https://user-images.githubusercontent.com/36160389/140775075-ed694a25-202f-4d2f-8c17-363001c0b9de.png)

![Screenshot from 2021-11-08 12-55-52](https://user-images.githubusercontent.com/36160389/140774986-ff85ae19-58ce-42de-b694-4221c075ab30.png)

## Expected

Same behavior as pytorch 1.8.1

## Environment

```
PyTorch version: 1.10.0
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Arch Linux (x86_64)
GCC version: (GCC) 11.1.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.33

Python version: 3.8.12 (default, Oct 12 2021, 13:49:34)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.14.16-arch1-1-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: NVIDIA GeForce GTX 1080
Nvidia driver version: 495.44
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.21.2
[pip3] torch==1.10.0
[pip3] torchvision==0.11.1
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               11.3.1               h2bc3f7f_2  
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] mkl                       2021.4.0           h06a4308_640  
[conda] mkl-service               2.4.0            py38h7f8727e_0  
[conda] mkl_fft                   1.3.1            py38hd3c417c_0  
[conda] mkl_random                1.2.2            py38h51133e4_0  
[conda] numpy                     1.21.2           py38h20f2e39_0  
[conda] numpy-base                1.21.2           py38h79a1101_0  
[conda] pytorch                   1.10.0          py3.8_cuda11.3_cudnn8.2.0_0    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchvision               0.11.1               py38_cu113    pytorch
```

 - PyTorch Version (e.g., 1.0): 1.10
 - OS (e.g., Linux): Linux
 - How you installed PyTorch (`conda`, `pip`, source): conda
 - Build command you used (if compiling from source):
 - Python version: 3.8
 - CUDA/cuDNN version:  11.3/8.2.0
 - GPU models and configuration:
 - Any other relevant information:

**Comment by thiagocrepaldi** [comment]
I could repro this. Indeed the `output` shape changes from `[batch,3,16,16]` to `[batch,Gatheroutput_dim_1,Gatheroutput_dim_2,Gatheroutput_dim_3]` when pytorch is upgraded from 1.8.2 to 1.10 or newer

**Comment by thiagocrepaldi** [comment]
@jiafatom Is this behavior expected when `dynamic_axes=None`?

**Comment by thiagocrepaldi** [comment]
I noticed that if you removed the `dynamic_axes` argument from `export()`, output shape will be `[2,3,16,16]`

**Comment by jiafatom** [comment]
If we don't set `dynamic_axes`, then the input shape is static, and the output shape is also static `[2,3,16,16]`, which is expected. When `dynamic_axes` set, current onnx shape inference cannot give the correct result for `Gather`, therefore it gives the output the dynamic shape. So we need either fix shape inference in onnx repo or make changes in exporter.

**Comment by garymm** [comment]
Just reproduced from tip of onnx_ms_1.
This is tracked internally at Microsoft by <https://msdata.visualstudio.com/Vienna/_workitems/edit/1524570>

**Comment by gnloch** [comment]
Is there any update on a fix for this issue? I tested on pytorch 1.10.2 and the bug still happens.

**Comment by BowenBao** [comment]
@gnloch created #73607 to fix this.
